mempool/src/shared_mempool/priority.rs

```diff
@@ -0,0 +1,728 @@
+    peer_comparator: PrioritizedPeersComparator,
+
+    // Whether ping latencies were observed for all peers
+    observed_all_ping_latencies: bool,
+
+    // The last time peer priorities were updated
+    last_peer_priority_update: Option<Instant>,
+
+    // The time service used to fetch timestamps
+    time_service: TimeService,
+}
+
+impl PrioritizedPeersState {
+    pub fn new(mempool_config: MempoolConfig, time_service: TimeService) -> Self {
+        Self {
+            mempool_config,
+            prioritized_peers: Arc::new(RwLock::new(Vec::new())),
+            peer_comparator: PrioritizedPeersComparator::new(),
+            observed_all_ping_latencies: false,
+            last_peer_priority_update: None,
+            time_service,
+        }
+    }
+
+    /// Returns the priority of the given peer. The lower the
+    /// value, the higher the priority.
+    pub fn get_peer_priority(&self, peer_network_id: &PeerNetworkId) -> usize {
+        let prioritized_peers = self.prioritized_peers.read();
+        prioritized_peers
+            .iter()
+            .find_position(|peer| *peer == peer_network_id)
+            .map_or(usize::MAX, |(position, _)| position)
+    }
+
+    /// Returns true iff the prioritized peers list is ready for another update.
+    /// This is based on the last time the prioritized peers were updated, and if
+    /// ping latencies were observed for all peers in the last update.
+    pub fn ready_for_update(&self, peers_changed: bool) -> bool {
+        // If our peers have changed, or we haven't observed ping latencies
+        // for all peers yet, we should update the prioritized peers again.
+        if peers_changed || !self.observed_all_ping_latencies {
+            return true;
+        }
+
+        // Otherwise, we should only update if enough time has passed since the last update
+        match self.last_peer_priority_update {
+            None => true, // We haven't updated yet
+            Some(last_update) => {
+                let duration_since_update = self.time_service.now().duration_since(last_update);
+                let update_interval_secs = self
+                    .mempool_config
+                    .shared_mempool_priority_update_interval_secs;
+                duration_since_update.as_secs() > update_interval_secs
+            },
+        }
+    }
+
+    /// Sorts the given peers by priority using the prioritized peer comparator.
+    /// The peers are sorted in descending order (i.e., higher values are prioritized).
+    fn sort_peers_by_priority(
+        &self,
+        peers_and_metadata: &[(PeerNetworkId, Option<PeerMonitoringMetadata>)],
+    ) -> Vec<PeerNetworkId> {
+        peers_and_metadata
+            .iter()
+            .sorted_by(|peer_a, peer_b| {
+                let ordering = &self.peer_comparator.compare(peer_a, peer_b);
+                ordering.reverse() // Prioritize higher values (i.e., sorted by descending order)
```

Comment:
@JoshLind - May be I am missing something but this seems to be returning the list in ascending order of priority - because the `peer_comparator` is returning the highest priority first?

Reply:
It's because `sorted_by` sorts in [ascending order](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.sorted_by), but we want high priority peers to be at the beginning of the list (i.e., descending order, as mempool selects peers from the front of the list). So, the easiest way to do that is reverse the ordering here 😄

Reply:
(We can also do this by flipping peer_a and peer_b, if that's clearer 🤔)

Reply:
Thanks - this makes sense to me.